### ISSUE TITLE
Feature - Enable format options for typescript (and others?)

### ISSUE THREAD

**Issue by SimplyKnownAsG** [ycm-core/ycmd#1379]
The java completer was given the ability to modify the `'format'` configuration [1] as a result of an issue in YouCompleteMe [2].

I'd like to be able to use a similar solution using `def Settings( **kwargs ):` from the `ycm_extra_conf`. The typescript completer has some thoughts [3] on enabling the other features, including loading from `tsconfig.json`.

[1] https://github.com/ycm-core/ycmd/pull/1339
[2] https://github.com/ycm-core/YouCompleteMe/issues/3505
[3] https://github.com/ycm-core/ycmd/blob/master/ycmd/completers/typescript/typescript_completer.py#L862

**Comment by bstaletic** [comment]
#1339 actually works for all LSP completers, which only leaves out tsserver. Python and C# currently don't support `Format` subcommand. C# completer could offer it, but it currently can't support formatting options - [there are none](https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Abstractions/Models/v1/CodeFormat/CodeFormatRequest.cs).

TSServer does support formatting options - https://github.com/Microsoft/TypeScript/blob/master/lib/protocol.d.ts#L2282-L2308

The following patch should enable formatting options, but it's completely untested and does no error checking:

```diff
diff --git a/ycmd/completers/typescript/typescript_completer.py b/ycmd/completers/typescript/typescript_completer.py
index c3666fea..0137f573 100644
--- a/ycmd/completers/typescript/typescript_completer.py
+++ b/ycmd/completers/typescript/typescript_completer.py
@@ -37,6 +37,7 @@ from ycmd import responses
 from ycmd import utils
 from ycmd.completers.completer import Completer
 from ycmd.completers.completer_utils import GetFileLines
+from ycmd.extra_conf_store import ModuleForSourceFile
 from ycmd.utils import LOGGER, re

 SERVER_NOT_RUNNING_MESSAGE = 'TSServer is not running.'
@@ -875,13 +876,21 @@ class TypeScriptCompleter( Completer ):
     # options, which is already adopted by a number of clients, would be to read
     # the "formatOptions" field in the tsconfig.json file.
     options = request_data[ 'options' ]
+    format_options = {
+      'tabSize': options.pop( 'tab_size' ),
+      'indentSize': options.pop( 'tab_size' )
+      'convertTabsToSpaces': options.pop( 'insert_spaces' )
+    }
+    module = ModuleForSourceFile( filepath )
+    settings = module.Settings(
+      language = 'typescript',
+      filename = filepath,
+      client_data = request_data[ 'extra_conf_data' ]
+    )
+    format_options.update( settings.get( 'format_options', {} ) )
     self._SendRequest( 'configure', {
       'file': filepath,
-      'formatOptions': {
-        'tabSize': options[ 'tab_size' ],
-        'indentSize': options[ 'tab_size' ],
-        'convertTabsToSpaces': options[ 'insert_spaces' ],
-      }
+      'formatOptions': format_options
     } )

     response = self._SendRequest( 'format',
```

**Comment by SimplyKnownAsG** [comment]
Good to know about `tsserver` being the outlier. This is what I have got: https://github.com/SimplyKnownAsG/ycmd/commit/349af582adc9eb515c43a22e7bb0a1aea4e7bc40

I'm happy to create a pull request. Should I do that? I'll need to spend some time figuring out the testing.

**Comment by bstaletic** [comment]
Oh, that looks way better. One thing I'm not sure about is that `def Language: return 'typescript'`, as that same completer is also used for javascript and typescriptreact.

The `Lanugage()` so far could return a hardcoded string, because all LSP completers except clangd, which historically used `'c-family'` as the language and lumped together C/C++/Objective-C/Objective-C++/CUDA. If we pass `request_data` through to `Language()` we can use `request_data[ 'filetypes' ][ 0 ]` for the typescript.

> I'm happy to create a pull request. Should I do that?

That would be great.

> I'll need to spend some time figuring out the testing.

#1339 has a good example of how a test for this could look like. If you need help, you can ask questions in [gitter](https://gitter.im/Valloric/ycmd).

**Comment by bstaletic** [comment]
No tests yet, but I did make a branch for this. https://github.com/bstaletic/ycmd/tree/ts-format

Wiring `request_data` through to `Language()` was much more intrusive than I thought it would be.

**Comment by bstaletic** [comment]
@SimplyKnownAsG Once again, thanks for the patch. This was implemented in #1450
